src/pass/llvmgen.cpp

```diff
@@ -457,7 +504,25 @@ void LLVMGen::visit(Func& func)
     auto entry_bb = BasicBlock::Create(llctx(), "entry", fn);
 
     builder()->SetInsertPoint(entry_bb);
-    builder()->CreateRet(eval(func.output));
+
+    auto output = eval(func.output);
+    if (func.is_kernel) {
+        builder()->CreateRetVoid();
+
+        fn->setCallingConv(llvm::CallingConv::PTX_Kernel);
+        llvm::NamedMDNode* MD =
+            llmod()->getOrInsertNamedMetadata("nvvm.annotations");
+
+        std::vector<llvm::Metadata*> MDVals;
+        MDVals.push_back(llvm::ValueAsMetadata::get(fn));
+        MDVals.push_back(llvm::MDString::get(llctx(), "kernel"));
+        MDVals.push_back(llvm::ConstantAsMetadata::get(
+            llvm::ConstantInt::get(llvm::Type::getInt32Ty(llctx()), 1)));
+
+        MD->addOperand(llvm::MDNode::get(llctx(), MDVals));
```

Comment:
I would wrap this part of the code under and #ifdef checking if CUDA is enabled.

Reply:
otherwise, I believe the compilation would fail if someone builds the code without cuda